engineering-plans/0004-subgraph-grafting.md

```diff
@@ -0,0 +1,143 @@
+# PLAN-0004: Subgraph Grafting
+
+<dl>
+  <dt>Author</dt>
+  <dd>David Lutterkort</dd>
+
+  <dt>Implements</dt>
+  <dd>No RFC</dd>
+
+  <dt>Engineering Plan pull request</dt>
+  <dd><a href="URL">URL</a></dd>
+
+  <dt>Date of submission</dt>
+  <dd>2020-03-17</dd>
+
+  <dt>Date of approval</dt>
+  <dd>YYYY-MM-DD</dd>
+
+  <dt>Approved by</dt>
+  <dd>First Person, Second Person</dd>
+</dl>
+
+## Contents
+
+<!-- toc -->
+
+## Summary
+
+This feature makes it possible to resume indexing of a failed subgraph by
+creating a new subgraph that uses the failed subgraph's data up to a block
+that the subgraph writer deems reliable, and continue indexing using the
+new subgraph's mappings. We call this operation *grafting* subgraphs.
+
+## Subgraph manifest changes
+
+The user-facing control for this feature will be a new optional
+`deployment` field in the subgraph's manifest. It can take one of three
+forms, and defaults to `create`:
+
+    ---
+    description: My really very good subgraph
+    # Option 1: (default) The deployment starts with no data in the database
+    deployment:
+      mode: create
+    # Option 2: (copy) The deployment starts with a copy of the data from
+    # the given subgraph at the given block
+    deployment:
+      mode: copy
+      subgraph: Qm...
+      block: 123456
+    # Option 3: (reuse) The deployment reuses the data of an existing subgraph
+    # in place. This mode implicitly deletes the source deployment, and any
+    # subgraph version using the source deployment will be switched to use the
+    # new deployment
+    deployment:
+      mode: reuse
+      subgraph: Qm...
+      block: 123456
+
+- For `copy` and `reuse` it is a deploy-time error if the source subgraph
+  does not exist or has not indexed the given `block` yet
+- To start with, for grafting, the source subgraph and the grafted subgraph
+  must have identical GraphQL schemas. We can expand/relax this requirement
```

Comment:
For this to be useful we may need to support a limited amount of schema changes at the outset - like converting a non-nullable field to a nullable field.